### ISSUE TITLE
Reported count of out-of-order samples can be misleading on edge cases

### ISSUE THREAD

**Issue by pracucci** [grafana/mimir#1711]
Each ingester tracks the number of out-of-order samples using this metric:
```
cortex_discarded_samples_total{reason="sample-out-of-order",user="<tenant-id>"}
```

The tracking of out-of-order samples is local in each ingester. Distributors replica each series to 3 ways, so under some edge conditions a write request could succeed if the out-of-order samples happen in only 1 out of the 3 ingesters (replication factor = 3, quorum = 2).

The edge case I'm referring to is:
- Prometheus (sender) is catching up after a short network outage
- One ingester is overloaded and slow, so samples in new write requests could be committed before samples in old requests are committed

To reproduce it I've used the local dev environment with the following changes:
- Reduced Prometheus scrape interval to 1s to simulate a Prometheus catching up
- Slow down 1 out of 2 requests only in 1 of the 3 ingesters

The code I've used is [here](https://github.com/grafana/mimir/compare/reproduce-out-of-order-samples-on-slow-ingester). Keep it running for few seconds and you see the tracking of out-of-order samples even if, from the Prometheus perspective, all requests succeed:
```
cortex_discarded_samples_total{reason="sample-out-of-order",user="anonymous"} 12819
```

> can be misleading

I said "misleading" because we're used to think about the tracking of out-of-order samples as the case samples have been written out of order from the source, but this edge case shows it could be tracked even if the source legitimately send all samples in order, in different requests.

**Comment by Dieterbe** [comment]
>  so samples in new write requests could be committed before samples in old requests are committed

does this take place within 1 ingester then? so, the root issue here is that within a single ingester process, if request B arrives after A has arrived, B might be committed before A?

**Comment by pracucci** [comment]
> does this take place within 1 ingester then?

Yes. Only ingesters can track out-of-order samples count. Distributors are not aware of it.

> so, the root issue here is that within a single ingester process, if request B arrives after A has arrived, B might be "processed" before A?

Yes.
